Initial changes for saving product images upload statuses in UserDefaults

pmusolino opened the pull request:
## Description

Part 1 of the changes needed for product image uploads in the background.

- **Purpose**: Implement initial changes to save product image upload statuses in `UserDefaults`. (The actual save will happen in another PR) Ref. pe5sF9-3Uc
- **Why**: 
  - Required for background image uploads.
  - Ensure a stored queue of image statuses persists when the app is in the background.
  - Queue includes `siteID` and `productID`.
- **Background**:
  - `URLSession` background configuration doesn't have access to `ServiceLocator` or similar classes.
  - System processes every element in scheduled `URLSession` upload tasks.
  - A queue must function even when the app isn't running.
- **Implementation**:
  - Introduces `ProductImagesUserDefaultsStatuses` for storing statuses in `UserDefaults`.
  - Not currently used; will be implemented in a future PR.
  - Suggest skipping the review of `ProductImagesUserDefaultsStatuses` due to significant changes in a subsequent PR.
- **Additional Changes**: 
  - More changes than expected due to file relocations.

#### Questions that may arise:

- **Why use `siteID` and `productID` in `ProductImageStatus`?**:
  - Background `URLSession` doesn't allow request updates during execution or scheduling. If the app goes in background, we do not have any power on it.
  - After user actions like pressing `save`, updating requests isn't feasible.
  - Then, Storing IDs helps check statuses from `UserDefaults` and send subsequent requests at the end of every future upload request.
  - Facilitates progress recovery and error management upon app re-launch from the background.

## Main changes applied in this PR:

- **File Movements**:
  - Moved `ProductImageStatus` to the Networking layer for background `URLSession` access.
  - Moved `ProductOrVariationID` enum to Networking layer for consistency.
- **Enhancements**:
  - `ProductImageStatus` and `ProductImageAssetType` now conform to `Codable` for `UserDefaults` storage.
  - `ProductImageStatus` enum cases (`uploading`, `remote`, `uploadFailure`) now accept `siteID` and `productID`.
- **New Implementation**:
  - `ProductImagesUserDefaultsStatuses` for `ProductImageStatus` storage.
  - Future changes expected, with unit tests to be added.
- **Code Updates**:
  - Updated image upload classes to handle `siteID` and `productID`.
  - Updated related unit tests.

## Testing information

- **Scenarios Tested**:
  - Uploading multiple images in a new product, both before and after publishing.
  - Uploading multiple images in an existing product, both before and after saving.
  - Selecting and saving new images from the WordPress media library.
  - Ensuring image re-ordering functionality.
  - Uploading images to product variations and saving them before/after upload completion.
  - Managing image upload errors effectively.


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Enhanced image upload handling to provide more robust status tracking and improved reliability.
- **Refactor**
	- Consolidated image management logic to ensure consistent state updates and streamlined user interactions.
- **Tests**
	- Expanded testing of image upload flows and status updates, ensuring a smoother experience when managing product images.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pmusolino committed:
refactor: moved `ProductImageStatus` into Networking layer and related changes

pmusolino committed:
Add Codable conformance and new file for storing product image statuses in UserDefaults
- Modify `ProductImageStatus.swift` to add Codable conformance and additional properties
- Create `ProductImagesUserDefaultsStatuses.swift` for managing product image statuses in User Defaults
- Move `ProductOrVariationID` in a separate file to define a type-safe identifier for products and variations
- Update `Model.swift` to include `ProductOrVariationID` typealias

pmusolino committed:
fix: spaces in `ProductImageStatus`

pmusolino committed:
Add siteID and productID to ProductImageStatus
- Update `ProductImagesCollectionViewDataSource.swift` to include siteID and productID in `ProductImageStatus` cases.
- Modify `ProductImagesHeaderTableViewCell.swift` to handle additional parameters in `ProductImageStatus`.
- Adjust `Product+Media.swift` to pass siteID and productID when mapping images to `ProductImageStatus`.
- Change `ProductImageActionHandler.swift` to incorporate siteID and productID in `ProductImageStatus` handling and updating.
- Update `ProductImageStatus+Extension.swift` to manage new parameters in `ProductImageStatus`.
- Amend `ProductImagesCollectionViewController.swift` to handle siteID and productID in status cases.
- Modify `ProductImagesSaver.swift` to use siteID and productID in status comparison and updates.
- Adjust `ProductImagesViewController.swift` to manage `ProductImageStatus` with siteID and productID.

pmusolino committed:
fix: when a product is published and `updateProductID(_:)` is called, all statuses without the correct value are now updated. This ensures that even if the statuses were initially created with a nil or outdated productID, they will be updated to the new remoteProductID.

pmusolino committed:
- Update `ProductImageStatus.swift` to make `productID` optional in `uploading` and `uploadFailure` cases
- Modify `ProductImagesUserDefaultsStatuses.swift` to handle optional `productID` in filtering statuses

pmusolino committed:
update: unit tests regarding product image upload classes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
update: removed superfluous comment

pmusolino committed:
fix: product image variation management, which allow maximum one image for variations.

pmusolino committed:
update: release-notes 21.8

pmusolino committed:
fix: failing unit tests

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults

pmusolino committed:
update: release-notes

pmusolino committed:
Merge commit 'a6877f9a55e964ace617c9e82a8b8897db842c53' into feat/save-product-image-upload-statuses-in-user-defaults

pmusolino committed:
- Update `ProductImageStatus.swift` to use non-optional `ProductOrVariationID` in `uploading` and `uploadFailure` cases
- Update unit tests

pmusolino committed:
fix: test_hasUnsavedChangesOnImages_stays_false_after_uploading_and_saving_successfully

pmusolino committed:
refactor: Set up the observer first to wait for the status change to pending upload, then trigger the media upload after adding the observer.

pmusolino committed:
Fix: asset state transition issue by updating .phAsset comparison logic. Instead of direct PHAsset comparison, compare localIdentifier to ensure accurate detection and state updates from "uploading" to "remote." This resolves the issue where hasUnsavedChangesOnImages(...) incorrectly returns true after saving, fixing related test failures.

pmusolino committed:
update: explicit injection of mock store dependencies in tests to avoid default `ServiceLocator` usage.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Cells/Product Images/ProductImagesCollectionViewDataSource.swift:
```diff
@@ -51,16 +51,16 @@ private extension ProductImagesCollectionViewDataSource {
+        case .remote(let image, let siteID, let productID):
             configureRemoteImageCell(cell, productImage: image, isFirstImage: isFirstImage)
-        case .uploading(let asset):
+        case .uploading(let asset, let siteID, let productID):
```
Nit: please omit the unused `siteID` and `productID` here to avoid confusion.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImageActionHandler.swift:
```diff
@@ -173,8 +175,15 @@ final class ProductImageActionHandler: ProductImageActionHandlerProtocol {
-            productImageStatuses = [.uploading(asset: asset)] + self.productImageStatuses
+            let uploadingStatus = ProductImageStatus.uploading(asset: asset, siteID: self.siteID, productID: self.productOrVariationID)
+
+            // If the product is a variation, substitute the existing status with the new uploading status.
+            // Otherwise, if a standard product, append a new status.
+            if case .variation = self.productOrVariationID {
+                self.productImageStatuses = [uploadingStatus]
+            } else {
+                self.productImageStatuses = [uploadingStatus] + self.productImageStatuses
+            }
```
While this check is correct as we allow only one image for variations, the logic should not be part of `ProductImageActionhandler` IMO. The UI should already take care of disabling uploading more than one image for variations. I'd suggest restoring the previous logic.

pmusolino replied:
I think this is a safeguard that makes sense to keep. The `ProductImageActionHandler` is one of the central pieces that manages and tracks image upload statuses. Even if the UI intends to allow only one image for variations, the handler is the final arbiter. And in fact, reverting the code generate a crash because the UI is not managing in a proper way the image statuses, replaing the current images showed. This means that if you upload an image in a variation, save it, then you try to replace it, you can expect a crash because in the end the images will be two (the remote one + the new one).

itsmeichigo replied:
Are you suggesting that the previous logic is causing a crash on `trunk`? I tried as you suggested but couldn't reproduce the crash. 

I think the crash should not happen because when you replace an image, the remote image is discarded from the status list.

itsmeichigo replied:
Tagging @pmusolino in case this message got lost is the depth of discussions here. Could you clarify my last question above please?

pmusolino replied:
@itsmeichigo No, I meant that given how the status search now works with siteID and productID, this looks like a necessary change in this PR. In trunk, everything should work as before.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImageActionHandler.swift:
```diff
@@ -295,9 +314,9 @@ private extension ProductImageActionHandler {
-            case .uploading(let uploadingAsset):
-                return uploadingAsset == asset
-            case let .uploadFailure(failedAsset, _):
+            case .uploading(let uploadingAsset, let siteID, let productID):
+                return uploadingAsset == asset && siteID == self.siteID && productID == self.productOrVariationID
+            case let .uploadFailure(failedAsset, _, _, _):
```
Any reason why we need to compare site ID and product ID for uploading status but not the failure status?

pmusolino replied:
Good catch, updated 👍

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImageStatus+Extension.swift:
```diff
@@ -1,49 +1,11 @@
         compactMap { status in
             switch status {
-            case .remote(let productImage):
+            case .remote(let productImage, let siteID, let productID):
```
Nit: please omit the unused params.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImageStatus+Extension.swift:
```diff
@@ -86,7 +48,7 @@ extension ProductImageStatus {
     var dragItemIdentifier: String {
         switch self {
-        case .uploading(let asset):
+        case .uploading(let asset, let siteID, let productID):
```
Nit: please omit the unused params.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImageStatus+Extension.swift:
```diff
@@ -95,7 +57,7 @@ extension ProductImageStatus {
         case .uploadFailure:
             return UUID().uuidString
-        case .remote(let image):
+        case .remote(let image, let siteID, let productID):
```
Nit: please omit the unused params.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesCollectionViewController.swift:
```diff
@@ -186,7 +186,7 @@ extension ProductImagesCollectionViewController {
         case .remote:
             break
-        case let .uploadFailure(asset, error):
+        case let .uploadFailure(asset, error, siteID, productID):
```
Nit: please omit the unused params.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesViewController.swift:
```diff
@@ -26,7 +26,7 @@ final class ProductImagesViewController: UIViewController {
         return productImageStatuses.compactMap { status in
             switch status {
-            case .remote(let productImage):
+            case .remote(let productImage, let siteID, let productID):
```
Nit: please omit the unused params.

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageUploaderTests.swift:
```diff
@@ -6,22 +6,25 @@ import Yosemite
+                                                 imagesProductIDUpdater: mockProductIDUpdater)
         let actionHandler = imageUploader.actionHandler(key: .init(siteID: siteID,
-                                                                   productOrVariationID: .product(id: productID),
+                                                                   productOrVariationID: .product(id: productID.id),
```
We can just use `productID` directly here right? Same comment for the changes below.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImageStatus.swift:
```diff
@@ -0,0 +1,177 @@
+/// The status of a Product image.
+///
+
+public enum ProductImageStatus: Equatable, Codable {
```
Since you added Codable conformance to `ProductImageStatus`, please also add tests to confirm that both the encoding and decoding work as expected.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImageStatus.swift:
```diff
@@ -0,0 +1,177 @@
+}
+
+/// The type of product image asset.
+public enum ProductImageAssetType: Equatable, Codable {
```
Since you added Codable conformance to `ProductImageAssetType `, please also add tests to confirm that both the encoding and decoding work as expected.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductOrVariationID.swift:
```diff
@@ -0,0 +1,54 @@
+/// Identifiable data about a product or product variation.
+public enum ProductOrVariationID: Equatable, Hashable, Codable {
```
Since you added Codable conformance to `ProductOrVariationID `, please also add tests to confirm that both the encoding and decoding work as expected.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImagesUserDefaultsStatuses.swift:
```diff
@@ -0,0 +1,99 @@
+/// Save product image upload statuses in User Defaults.
+/// This class is declared in the Networking layer because it will also be accessed by the background URLSession operations.
+///
+final class ProductImagesUserDefaultsStatuses {
```
Please consider renaming this to `ProductImageStatusStorage` and injecting user defaults into its initializer for testability.

pmusolino replied:
This file has been extensively revised in a future PR, where UserDefaults is also injected for testability. Anyway, the suggested name seems interesting to me. Since this file has been thoroughly revisited in future PRs, I will rename it there. 👍

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImageStatus.swift:
```diff
@@ -0,0 +1,177 @@
+        case "uploadFailure":
+            let asset = try container.decode(ProductImageAssetType.self, forKey: .asset)
+            let errorMessage = try container.decode(String.self, forKey: .error)
+            let error = NSError(domain: "ProductImageStatus", code: 0, userInfo: [NSLocalizedDescriptionKey: errorMessage])
```
Looks like we'll lose information about the error here. The information is necessary because we're showing details about the upload failures. Please also retain the domain and error code while encoding/decoding.

pmusolino replied:
This has been handled in a separate PR, but I'm applying the changes from that PR in this one. 👍

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImageStatus.swift:
```diff
@@ -0,0 +1,177 @@
+            try container.encode("uploading", forKey: .type)
+            try container.encode(asset, forKey: .asset)
+            try container.encode(siteID, forKey: .siteID)
+            try container.encodeIfPresent(productID, forKey: .productID)
```
Any reason why `encodeIfPresent` is necessary? I think `productID` is non-optional here.

pmusolino replied:
Correct. In fact, in a following PR, this has been changed. Like above, I'm applying the changes from that PR in this one.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImageStatus.swift:
```diff
@@ -0,0 +1,177 @@
+            let errorMessage = (error as NSError).localizedDescription
+            try container.encode(errorMessage, forKey: .error)
+            try container.encode(siteID, forKey: .siteID)
+            try container.encodeIfPresent(productID, forKey: .productID)
```
Same comment about `encodeIfPresent` as above.

itsmeichigo commented on Networking/Networking/ProductImageInBackground/ProductImagesUserDefaultsStatuses.swift:
```diff
@@ -0,0 +1,99 @@
+                if let filterProductID = productID {
+                    return sID == siteID && pID == filterProductID
+                } else {
+                    return sID == siteID && pID == nil
```
Product ID is non-optional in image statuses so this check will likely fail.

pmusolino replied:
As mentioned in the PR description, I suggested not reviewing this file since it is not used and, above all, it has been deeply modified in a future PR (specifically in [this one](https://github.com/woocommerce/woocommerce-ios/pull/15246/files), which will be split into multiple PRs due to its size). This code no longer exists there 👍 so you can ignore it for the moment.

itsmeichigo replied:
If this file is not ready, please do not commit it into a PR like this. It would be hard for me to approve the PR because not everything is working and I cannot guarantee the quality of the code to be merged.

itsmeichigo reviewed (commented):
Hi Paolo, this PR is really large so it was hard to review. I'm leaving all my comments here first before testing the app. 

Next time, it'd be great to split PRs into smaller pieces and have thorough tests for each change rather than wrapping everything in one go. For example:
- Part 1: Create copies of `ProductImageStatus`, `ProductOrVariationID` in the Networking layer and add tests for the codable conformance.
- Part 2: Add the storage for product image status and unit test it.
- Part 3: Replace the networking types in the UI layer.

pmusolino committed:
update: omit the unused siteID and productID to avoid confusion

pmusolino committed:
Update method parameter from `.product(id: productID.id)` to `.productID` in `ProductImageUploaderTests.swift`

pmusolino committed:
Enhance error handling in ProductImageStatus

- Add new cases for `errorDomain`, `errorCode`, and `errorUserInfo` in `ProductImageStatus.swift`.
- Modify decoding logic to handle detailed error information.
- Update encoding logic to include `errorDomain`, `errorCode`, and `errorUserInfo`.
- Change `productID` to be always encoded in `uploading` and `uploadFailure` cases.

pmusolino committed:
Update the upload failure condition to include siteID and productID comparison in ProductImageActionHandler

pmusolino committed:
fix: passing siteID and productID in uploadFailure

pmusolino committed:
fix: typo

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults

pmusolino commented on Networking/Networking/ProductImageInBackground/ProductImagesUserDefaultsStatuses.swift:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    static func setAllStatuses(_ statuses: [ProductImageStatus], for siteID: Int64, productID: ProductOrVariationID?) {
+        // Merge with existing, removing any old statuses for this site/product
+        var all = getAllStatuses().filter { st in
+            switch st {
+            case .remote(_, let sID, let pID):
+                if let filterProductID = productID {
+                    return !(sID == siteID && pID == filterProductID)
+                } else {
+                    return true
+                }
+            case .uploading(_, let sID, let pID),
+                 .uploadFailure(_, _, let sID, let pID):
+                if let filterProductID = productID {
+                    return !(sID == siteID && pID == filterProductID)
+                } else {
+                    return !(sID == siteID && pID == nil)
+                }
+            }
+        }
+        all.append(contentsOf: statuses)
+        saveAllStatuses(all)
+    }
```
This does not apply, also because this class has been completely rewritten in another PR.

pmusolino committed:
- Modify `ProductImageActionHandlerTests.swift` to use `siteID` instead of a hardcoded value.
- Change `ProductImageUploaderTests.swift` to correctly assign `productID.id` to `remoteProductID`.

itsmeichigo reviewed (commented):
Hey Paolo, I tested the image upload and reordering on this PR and it seems to still work as before. However, it's really hard for me to approve this PR given that the file `ProductImagesUserDefaultsStatuses` is not ready. 

I understand that you are changing the file in a subsequent PR - how about removing it from this branch and adding it only to your subsequent PR? That way, we would not have to merge unfinished code.

pmusolino committed:
Update: Removed `ProductImagesUserDefaultsStatuses` since it has been extensively updated later on.

pmusolino commented:
Thank you @itsmeichigo  for your patience and review. I will split the next PRs into multiple PRs (I have at least 3 in the queue), because there is so much code, and I removed from this one what has been heavily modified in a future PR (in this case `ProductImagesUserDefaultsStatuses`). 6ea99b33e9873bd8b8a72dad3c465e964e9cc5c8

itsmeichigo reviewed (approved):
Thanks Paolo for the updates! I see that you're adding extra tests in a separate PR, so I'm approving this PR for now. Really appreciate your patience!

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults

pmusolino committed:
Merge branch 'trunk' into feat/save-product-image-upload-statuses-in-user-defaults
